src/test_memcachedstore.cpp

```diff
@@ -0,0 +1,509 @@
+}
+
+
+TYPED_TEST(SingleMemcachedStoreTest, SecondAddFails)
+{
+  Store::Status status;
+  const std::string data_in1 = "kermit";
+  const std::string data_in2 = "gonzo";
+  std::string data_out;
+  uint64_t cas;
+
+  status = this->_store->set_data(this->_table, this->_key, data_in1, 0, 300, DUMMY_TRAIL_ID);
+  EXPECT_EQ(status, Store::OK);
+
+  status = this->_store->get_data(this->_table, this->_key, data_out, cas, DUMMY_TRAIL_ID);
+  EXPECT_EQ(status, Store::OK);
+  EXPECT_EQ(data_out, data_in1);
+
+  status = this->_store->set_data(this->_table, this->_key, data_in2, 0, 300, DUMMY_TRAIL_ID);
+  EXPECT_EQ(status, Store::DATA_CONTENTION);
+
+  status = this->_store->get_data(this->_table, this->_key, data_out, cas, DUMMY_TRAIL_ID);
+  EXPECT_EQ(status, Store::OK);
+  EXPECT_EQ(data_out, data_in1);
+
+  status = this->_store->delete_data(this->_table, this->_key, DUMMY_TRAIL_ID);
+  EXPECT_EQ(status, Store::OK);
+}
+
+
+TYPED_TEST(SingleMemcachedStoreTest, KeyCanBeDeletedThenAdded)
+{
+  Store::Status status;
+  const std::string data_in1 = "kermit";
+  const std::string data_in2 = "gonzo";
+  std::string data_out;
+  uint64_t cas;
+
+  status = this->_store->set_data(this->_table, this->_key, data_in1, 0, 300, DUMMY_TRAIL_ID);
+  EXPECT_EQ(status, Store::OK);
+
+  status = this->_store->delete_data(this->_table, this->_key, DUMMY_TRAIL_ID);
+  EXPECT_EQ(status, Store::OK);
+
+  status = this->_store->set_data(this->_table, this->_key, data_in2, 0, 300, DUMMY_TRAIL_ID);
+  EXPECT_EQ(status, Store::OK);
+
+  status = this->_store->get_data(this->_table, this->_key, data_out, cas, DUMMY_TRAIL_ID);
+  EXPECT_EQ(status, Store::OK);
+  EXPECT_EQ(data_out, data_in2);
+
+  status = this->_store->delete_data(this->_table, this->_key, DUMMY_TRAIL_ID);
+  EXPECT_EQ(status, Store::OK);
+}
+
+//
+// MemcachedStore tests that only apply to when the store uses tombstone
+// records.
+//
+
+class MemcachedStoreTombstoneTest : public MemcachedTest
+{
+public:
+  MemcachedStore* _store;
+
+  virtual void SetUp()
+  {
+    MemcachedTest::SetUp();
+    _store = new MemcachedStore(false, new TombstoneConfig());
+  }
+
+  virtual void TearDown()
+  {
+    delete _store; _store = NULL;
+    MemcachedTest::TearDown();
+  }
+
+  static void SetUpTestCase()
+  {
+    MemcachedTest::SetUpTestCase();
+  }
+};
+
+TEST_F(MemcachedStoreTombstoneTest, TombstonesPreventSimpleAdds)
+{
+  Store::Status status;
+  memcached_return_t rc;
+  const std::string data_in1 = "kermit";
+  const std::string data_in2 = "gonzo";
+  std::string data_out;
+  uint64_t cas;
+
+  status = _store->set_data(_table, _key, data_in1, 0, 300, DUMMY_TRAIL_ID);
+  EXPECT_EQ(status, Store::OK);
+
+  status = _store->delete_data(_table, _key, DUMMY_TRAIL_ID);
+  EXPECT_EQ(status, Store::OK);
+
+  rc = simple_add(fqkey(), data_in2);
+  EXPECT_EQ(MEMCACHED_NOTSTORED, rc);
+
+  rc = simple_get(fqkey(), data_out, cas);
+  EXPECT_MEMCACHED_SUCCESS(rc, _memcached_client);
+  EXPECT_EQ(data_out, "");
```

Comment:
Fair point. I was trying to prove that the tombstone is not a normal record, but I have already "proved" that by the fact it can be overwritten with a 0 cas. I will remove this line.